src/core.c/array_slice.pm6

```diff
@@ -183,10 +183,513 @@ multi sub postcircumfix:<[ ]>( \SELF, Any:D \pos, Bool() :$v!, *%other ) is raw
 }
 
 # @a[@i]
-multi sub postcircumfix:<[ ]>( \SELF, Iterable:D \pos ) is raw {
-    nqp::iscont(pos)
-      ?? SELF.AT-POS(pos.Int)
-      !! POSITIONS(SELF, pos).map({ SELF[$_] }).eager.list;
+multi sub postcircumfix:<[ ]>(\SELF, Iterable:D \positions, *%_) is raw {
+    # MMD is not behaving itself so we do this by hand.
+    return postcircumfix:<[ ]>(SELF, positions.Int, |%_) 
+      if nqp::iscont(positions);
+
+    my $result  := nqp::create(IterationBuffer);
+    my $elems   := nqp::null;
+    my int $done;
+
+    # Helper sub for deleting elements, making sure that the total number
+    # of elements is fixed from *before* the first deletion, so that relative
+    # positions such as *-1 will continue to refer to the same position,
+    # even if the last element of an array was removed (which shortens the
+    # array).
+    sub delete(\pos) {
+        $elems := SELF.elems if nqp::isnull($elems);
+        SELF.DELETE-POS(pos)
+    }
+
+    # Get the dispatch index
+    my int $index;
+    if nqp::getattr(%_,Map,'$!storage') {
+        if nqp::istype(
+             (my $lookup := Rakudo::Internals.ADVERBS_TO_DISPATCH_INDEX(%_)),
+             X::Adverb
+           ) {
+            $lookup.what   = "slice";
+            $lookup.source = try { SELF.VAR.name } // SELF.^name;
+            return Failure.new($lookup);
+        }
+
+        # Good to go!
+        $index = $lookup;
+    }
+
+    my &accept := do if $index {
+
+        # Handlers for lazy positions and adverbs
+        if positions.is-lazy {
+               nqp::iseq_i($index,1) || nqp::iseq_i($index,2)
+              ?? -> \pos {    # lazy, :kv | :!kv
```

Comment:
Since the approach with a dispatch table is chosen, wouldn't it be faster to keep all the candidates in a local table and then pull `&accepts` from that table? I think it should be some performance gain comparing to the long chain of conditionals.

Reply:
I'm glad you ask this question.  And the sad answer is: this approach is chosen for a reason.

The reason is that every code block that is **accessed** in a block, will case a clone to be made of that block **and** of its closure.  And initializing it into an nqp::list counts as accessing.  If you have a dispatch table with 34 entries, that's an additional 68 allocations *before* you actually start to do anything.  And these allocations and closure taking completely wipe out **any** performance gain.

On the other hand, one could think of initializing the dispatch table at compile time.  But then they have taken the wrong closure.

To make dispatch tables like this work, the cloning of the block and closure taking should become lazy, I'd think.  Or provide some syntactic / pragma / nqp::code support to mark a block as intended to be cloned lazily.

@jnthn @niner

Reply:
`nqp::getstaticcode`? Pass `$result` as a parameter and get `$done` as a return value? Note that I checked out only few variants and don't know if any other lexical is used elsewhere.

Reply:
Possibly.  But I don't see it documented in https://github.com/Raku/nqp/blob/master/docs/ops.markdown ?  Is it documented somewhere else?  If not, maybe it should be documented ?

Reply:
Passing `$results` is not enough, `SELF` should then also be passed.  Which makes *3* parameters needing passing.  Which, if I recall correctly as I *have* tried that approach, basically kills performance again   :-(

Reply:
> Possibly. But I don't see it documented in https://github.com/Raku/nqp/blob/master/docs/ops.markdown ? Is it documented somewhere else? If not, maybe it should be documented ?

There're a lot of ops awaiting for documentation. Yet, `getstaticcode` is extensively used by `BOOTSTRAP`. I'd say it's implicitly documented. :) There is also one documented, `markcodestatic`. But I'm not really sure I understand it correctly. And if I am then it is only related to serialization.

> Passing `$results` is not enough, `SELF` should then also be passed. Which makes _3_ parameters needing passing. Which, if I recall correctly as I _have_ tried that approach, basically kills performance again :-(

What if the data to process is containerized in a class? I mean, it is really confusing that a long chain of `if`s works faster than anything else.

Reply:
Is there any way to use multi dispatch here? In Inline::Perl5 I had a similar dispatch chain. Since I try very hard to avoid any NQP there, I came up with this: https://github.com/niner/Inline-Perl5/commit/97ae4c7a3bcff892e7494e6f5824d973d282684d?branch=97ae4c7a3bcff892e7494e6f5824d973d282684d&diff=split and that made it faster and the code much nicer at the same time.

Reply:
@niner The previous implementation used multiple dispatch.  But since the adverbs are all passed as nameds, and the dispatcher gets really confused when having multiple required named parameters (let alone the maintainer), it quickly boils down to checking for one named parameter, and then looking at `%_` for additional named parameters.  Which totally sucks.  But as soon as you want to take all of that handling out of dispatch, you will need to handle the "no named parameters" case as well, as there is no way to dispatch on "no named arguments" if one of the candidates slurps all of the named arguments.

@vrurg ok, so making 2 x 34 classes from a role, and use a dispatch table to select which class to create the handler object with.  Hmmm...   That might work, but will bloat the core significantly, as each class comes with a lot of memory / disk use, no?

Reply:
> ok, so making 2 x 34 classes from a role, and use a dispatch table to select which class to create the handler object with. Hmmm... That might work, but will bloat the core significantly, as each class comes with a lot of memory / disk use, no?

No. I meant to use the table-based dispatch. Then make a single class hold all the parameters we may need and pass its instance as a single parameter to all subs. Something like:

```
my class DispatchParams {
    has $!SELF;
    has $!result;
    has $!pos;
}
my $dparam := nqp::create(DispatchParams);
nqp::bindattr($dparam, DispatchParams, '$!result', $result); # bind to the container
@table[$index].($dparam);

# inside the handler code:
nqp::getattr($dparam, DispatchParams, '$!result') = True; # should work because $!result =:= $result
```

This would eliminate the 3-parameter issue.

Hope I'm not saying anything stupid because I'm not fully awoken yet. :)

Reply:
I've taken @vrurg's suggestion and created a role with a class to implement the specific handling and run this in the branch.  The idea being that we would have a dispatch table of classes, and then just select the right class to do the work.  See code below.

Sadly, this approach for the most simple case, is about 2.5x as slow, basically making it perform as before.  But with a massive greater codegenned footprint, as this approach would require 68 consuming classes of the role to be made.

@vrurg is that what you had in mind?

```raku
use MONKEY;

my role PostcircumfixSlice {
    has $!result;
    has $!elems;
    has $!iterable;
    has int $!done;

    method !SET-SELF(\iterable) {
        $!result   := nqp::create(IterationBuffer);
        $!elems    := nqp::null;
        $!iterable := iterable;
        self
    }
    method new(\iterable) { nqp::create(self)!SET-SELF(iterable) }

    # The method that needs to be implemented by the consuming class
    method !accept(\pos) { ... }

    # Helper method for deleting elements, making sure that the total number
    # of elements is fixed from *before* the first deletion, so that relative
    # positions such as *-1 will continue to refer to the same position,
    # even if the last element of an array was removed (which shortens the
    # array).
    method !delete(\pos) {
        $!elems := $!iterable.elems if nqp::isnull($!elems);
        $!iterable.DELETE-POS(pos)
    }

    method slice(\iterator) {

        # Handle iterator in the generated positions: this will add a List
        # with the elements pointed to by the iterator to the result.  Because
        # these positions can also be non-Int, some trickery needs to be
        # done to allow this being called recursively.
        method !handle-iterator(\iterator) {

            # basically push the current result on a stack
            my int $mark = nqp::elems($!result);

            nqp::until(
              $!done || nqp::eqaddr((my \pos := iterator.pull-one),IterationEnd),
              nqp::if(
                nqp::istype(pos,Int),
                self!accept(pos),
                self!handle-nonInt(pos)
              )
            );

            # Take what was added and push it as a List
            if nqp::sub_i(nqp::elems($!result),$mark) -> int $added {
                my $buffer;
                if $mark {
                    $buffer := nqp::slice($!result,$mark,$added);
                    nqp::setelems($!result,$mark);
                }
                else {
                    $buffer := $!result;
                    $!result := nqp::create(IterationBuffer);
                }
                nqp::push($!result,$buffer.List);
            }
        }

        # Handle anything non-integer in the generated positions
        method !handle-nonInt(\pos) {
            nqp::istype(pos,Iterable)
              ?? nqp::iscont(pos)
                ?? self!accept(pos.Int)
                !! self!handle-iterator(pos.iterator)
              !! nqp::istype(pos,Callable)
                ?? nqp::istype(
                     (my $real := (pos)(
                       nqp::ifnull($!elems,$!elems := $!iterable.elems)
                     )),
                     Int
                   )
                  ?? self!accept($real)
                  !! self!handle-nonInt($real)
                !! nqp::istype(pos,Whatever)
                  ?? self!handle-iterator(
                       (^nqp::ifnull(
                           $!elems,
                           $!elems := $!iterable.elems
                         )).iterator
                     )
                  !! self!accept(pos.Int)
        }

        # Do the actual building of the result
        nqp::until(
          $!done || nqp::eqaddr((my \pos := iterator.pull-one),IterationEnd),
          nqp::if(
            nqp::istype(pos,Int),
            self!accept(pos),
            self!handle-nonInt(pos)
          )
        );

        $!result.List
    }
}

class Array::PostcircumfixSlice does PostcircumfixSlice {
    method !accept(\pos --> Nil) {
        nqp::push($!result,$!iterable.AT-POS(pos))
    }
}

# @a[@i]
sub foo(\SELF, Iterable:D \positions, *%_) is raw {
    return postcircumfix:<[ ]>(SELF, positions.Int, |%_) 
      if nqp::iscont(positions);

    # Get the dispatch index
    my int $index;
    if nqp::getattr(%_,Map,'$!storage') {
        if nqp::istype(
             (my $lookup := Rakudo::Internals.ADVERBS_TO_DISPATCH_INDEX(%_)),
             X::Adverb
           ) {
            $lookup.what   = "slice";
            $lookup.source = try { SELF.VAR.name } // SELF.^name;
            return Failure.new($lookup);
        }

        # Good to go!
        $index = $lookup;
    }
    Array::PostcircumfixSlice.new(SELF).slice(positions.iterator)
}

my @a = <a b c d e f g h i j>;

my $then;
my @i = 1,2,3;

$then = now;
@a[@i] for ^100000;
say now - $then;

$then = now;
foo(@a, @i) for ^100000;
say now - $then;
```

Reply:
Hmmm... looking at a profile of this, the top one is: `dispatch:<!>` at **28%**.  So apparently the private method dispatch is not being optimized.  And the `!accept` method is not being optimized in any way, shape or form (stays red in the profile).

Reply:
Ok, that's an interesting development: turning the whole thing into just a class (without the role business), makes this run **faster** than my current implementation.  Apparently, role composition is making a number of optimizations not work?